pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +136,511 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	if err != nil {
-		reqLogger.Errorf("Failed to update NodeMaintenance with \"Succeeded\" status. Error: %v", err)
-		return r.reconcileAndError(instance, err)
+func (r *ReconcileNodeMaintenance) doesLeaseExist(client client.Client, nodeName string) (int, error, *coordv1beta1.Lease) {
+	lease := &coordv1beta1.Lease{}
+
+	nName := apitypes.NamespacedName{Namespace: corev1.NamespaceNodeLease, Name: nodeName}
+
+	if err := client.Get(context.TODO(), nName, lease); err != nil {
+		if errors.IsNotFound(err) {
+			return LeaseStatusNotFound, nil, nil
+		}
+		return LeaseStatusFail, err, nil
 	}
 
-	return reconcile.Result{}, nil
-}
+	// check if the existing lease is currently active
+	leaseDeadline := (*lease.Spec.AcquireTime).Time
+	leaseDeadline.Add(time.Duration(*lease.Spec.LeaseDurationSeconds) * time.Second)
 
-func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
-	node, err := r.fetchNode(nodeName)
-	if err != nil {
-		return err
+	if leaseDeadline.After(time.Now()) {
+		// lease active
+
+		if LeaseHolderIdentity == *lease.Spec.HolderIdentity {
+			return LeaseStatusActiveOwnedByMe, nil, lease
+		} else {
+			return LeaseStatusActiveOwnedByDifferentHolder, nil, lease
+		}
 	}
 
-	// Uncordon the node
-	err = AddOrRemoveTaint(r.drainer.Client, node, false)
-	if err != nil {
+	// existing lease expired.
+	return LeaseStatusExpired, nil, nil
+}
+
+func (r *ReconcileNodeMaintenance) createOrUpdateLease(node *corev1.Node, RequestedMaintenanceWindowSeconds int32, lease *coordv1beta1.Lease) error {
+
+	nodeName := node.ObjectMeta.Name
+
+	if RequestedMaintenanceWindowSeconds < MinMaintenenceWindowSeconds {
+		err := fmt.Errorf("node %s maintenance window size too small", nodeName)
+		r.reqLogger.Errorf("%v", err)
 		return err
 	}
+	tmpDurationInSeconds := RequestedMaintenanceWindowSeconds
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       node.ObjectMeta.Name,
+		UID:        node.ObjectMeta.UID,
+	}
+	if lease == nil {
+
+		lease = &coordv1beta1.Lease{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            nodeName,
+				Namespace:       corev1.NamespaceNodeLease,
+				OwnerReferences: []metav1.OwnerReference{*owner},
+			},
+			Spec: coordv1beta1.LeaseSpec{
+				HolderIdentity:       &holderIdentity,
+				LeaseDurationSeconds: &tmpDurationInSeconds,
+				AcquireTime:          &metav1.MicroTime{Time: time.Now()},
+			},
+		}
+
+		if err := r.client.Create(context.TODO(), lease); err != nil {
+			r.reqLogger.Errorf("Failed to create lease. node %s error: %v", nodeName, err)
+			return err
+		}
+	} else {
+		lease.ObjectMeta.Name = nodeName
+		lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
+		lease.Spec.HolderIdentity = &holderIdentity
+		lease.Spec.LeaseDurationSeconds = &tmpDurationInSeconds
+		lease.Spec.AcquireTime = &metav1.MicroTime{Time: time.Now()}
+		lease.Spec.RenewTime = nil
```

Comment:
not updated anywhere

Reply:
i need to reset all existing fields when updating a lease (instead of deleting/creating it anew), therefore need to set RenewTime and LeaseTransition to nil; where am I wrong?